### Agent with Tools , With Memory (Trimming Messages , Summarization)

`Pros`

    . Any Question Regarding the Turtles ==> Turtles Tool
    . Any Question Other than this Uses Internet
    . Has Memory ==> Can be used with either persistence Memory or In Memory , this Notebook has an In Memory

`Cons`

    . No Control Flow , if the answer is not found in the database , it won't search the internet
    . Internet data is not saved yet , so we are still relying on Tavily

`The Agent is better than the chains because it can reason its working , ex : 3 photos for loggerhead , 2 photos for green turtle`

In [44]:
from langchain_core.messages import AIMessage, HumanMessage
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_openai import ChatOpenAI , OpenAIEmbeddings
from langchain.memory import ChatMessageHistory
from langchain_core.runnables.history import RunnableWithMessageHistory
from langchain_core.runnables import RunnablePassthrough
from langchain_community.vectorstores import FAISS
from langchain.tools.retriever import create_retriever_tool
from langchain_community.tools.tavily_search import TavilySearchResults
from langchain_community.tools import tool
from langchain.agents import AgentExecutor
from langchain.agents import create_tool_calling_agent
from langchain_groq import ChatGroq
from langchain import hub
import json
from langchain_deepseek import ChatDeepSeek
import requests
from langchain_ollama import ChatOllama
import random
from dotenv import load_dotenv

load_dotenv("../.env")

True

In [193]:
chat = ChatDeepSeek(
    model="deepseek-r1:14b",
    temperature=0,
    max_tokens=None,
    timeout=None,
    max_retries=2,
    api_key=DEEPSEEK_API_KEY,
    # other params...
)

In [209]:
chat = ChatOllama(model="deepseek-r1:14b")

In [201]:
chat = ChatOpenAI(model="gpt-3.5-turbo")

In [14]:
demo_ephemeral_chat_history = ChatMessageHistory()

In [211]:
prompt = ChatPromptTemplate.from_messages(
    [
        (
            "system",
            "You are a helpful assistant. Answer all questions to the best of your ability.",
        ),
        MessagesPlaceholder(variable_name="chat_history"),
        ("human", "{input}"),
    ]
)

chain = prompt | chat

In [16]:
def trim_messages(chain_input):
    stored_messages = demo_ephemeral_chat_history.messages
    if len(stored_messages) <= 2:
        return False

    demo_ephemeral_chat_history.clear()

    for message in stored_messages[-2:]:
        demo_ephemeral_chat_history.add_message(message)

    return True

In [213]:

chain_with_message_history = RunnableWithMessageHistory(
    chain,
    lambda session_id: demo_ephemeral_chat_history,
    input_messages_key="input",
    history_messages_key="chat_history",
)

### Trimmming

In [214]:
chain_with_trimming = (
    RunnablePassthrough.assign(messages_trimmed=trim_messages)
    | chain_with_message_history
)

In [215]:
config = {"configurable" : {"session_id" : "unused"}}

In [216]:
chain_with_trimming.invoke({"input" : "what was the first operation we did again?" } , config= config)

AIMessage(content='<think>\nOkay, so I\'m trying to figure out what the user meant by "what was the first operation we did again?" It\'s a bit unclear because it depends on the context they\'re referring to. Without more information, it\'s hard to pinpoint exactly what they\'re asking about.\n\nMaybe they\'re talking about a specific project or task we worked on together before. I should consider if there were any prior operations mentioned in our previous conversations. If so, recalling that context would be crucial to giving an accurate answer. However, since this is the first time they\'ve asked this question, it\'s possible they might be referring to something general.\n\nAnother angle could be that they\'re asking about a common first step in various processes. For example, in software development, the first operation might involve setting up the environment or initializing a project. In math, maybe the first operation is addition. But without knowing the specific field, it\'s har

In [21]:
demo_ephemeral_chat_history.messages

[HumanMessage(content='divide by 2', additional_kwargs={}, response_metadata={}),
 AIMessage(content='24 divided by 2 equals 12.', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 10, 'prompt_tokens': 54, 'total_tokens': 64, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_name': 'gpt-3.5-turbo-0125', 'system_fingerprint': None, 'finish_reason': 'stop', 'logprobs': None}, id='run-8c7f7c62-0280-45e5-8f49-3b68ff138440-0', usage_metadata={'input_tokens': 54, 'output_tokens': 10, 'total_tokens': 64, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 'reasoning': 0}}),
 HumanMessage(content='what was the first operation we did again?', additional_kwargs={}, response_metadata={}),
 AIMessage(content='The first operation we did was division.', addit

### Summarization

In [19]:
def summarize_messages(chain_input):
    stored_messages = demo_ephemeral_chat_history.messages
    if len(stored_messages) == 0:
        return False
    summarization_prompt = ChatPromptTemplate.from_messages(
        [
            MessagesPlaceholder(variable_name="chat_history"),
            (
                "user",
                "Distill the above chat messages into a single summary message. Include as many specific details as you can.",
            ),
        ]
    )
    summarization_chain = summarization_prompt | chat

    summary_message = summarization_chain.invoke({"chat_history": stored_messages})

    demo_ephemeral_chat_history.clear()

    demo_ephemeral_chat_history.add_message(summary_message)

    return True

In [20]:
chain_with_summarization = (
    RunnablePassthrough.assign(messages_summarized = summarize_messages)
    | chain_with_message_history
)

In [22]:
chain_with_summarization.invoke({"input" : "multiply the output by 5"} , 
                                config= config)

AIMessage(content='To multiply the output of 12 by 5, we simply multiply 12 by 5, which equals 60.', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 26, 'prompt_tokens': 67, 'total_tokens': 93, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_name': 'gpt-3.5-turbo-0125', 'system_fingerprint': None, 'finish_reason': 'stop', 'logprobs': None}, id='run-d3fa1730-77b2-4d35-84f1-40786c1838a9-0', usage_metadata={'input_tokens': 67, 'output_tokens': 26, 'total_tokens': 93, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 'reasoning': 0}})

In [23]:
demo_ephemeral_chat_history.messages

[AIMessage(content='In the conversation, we performed a division operation by dividing 24 by 2, which equals 12. The first operation we did was division.', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 31, 'prompt_tokens': 74, 'total_tokens': 105, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_name': 'gpt-3.5-turbo-0125', 'system_fingerprint': None, 'finish_reason': 'stop', 'logprobs': None}, id='run-62f19c95-910d-42ad-ba1e-5c3a19cee46e-0', usage_metadata={'input_tokens': 74, 'output_tokens': 31, 'total_tokens': 105, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 'reasoning': 0}}),
 HumanMessage(content='multiply the output by 5', additional_kwargs={}, response_metadata={}),
 AIMessage(content='To multiply the output of 12 by 5, we s

## Retreiver

In [45]:
vector_store = FAISS.load_local("../../Embeddings VS/OpenAIEmbed2" , embeddings=OpenAIEmbeddings() , allow_dangerous_deserialization=True)

In [46]:
retriever = vector_store.as_retriever(kwargs={"k":5})

In [47]:
retriever_tool = create_retriever_tool(
    retriever ,
    "Turtles_search",
    "Searches for information about the turtles , for any questions regarding the turtles you should use this tool! "
)

## Search Tool

In [48]:
search = TavilySearchResults()

## Photos Tool

In [49]:
def get_taxa_id(query):
    base_url = "https://api.inaturalist.org/v1/taxa"

    params = {"q": query}
    
    response = requests.get(base_url, params=params, timeout=10)
    response.raise_for_status()
    data = response.json() 
    
    results = data["results"]
    if results :
        return results[0]["id"]

    return {"error": "Please provide a valid query"}


def prepare_content(output):
    if isinstance(output, dict):
        # Serialize dictionary into JSON format
        return json.dumps(output, indent=2)
    return str(output)

from langchain.tools import tool
@tool
def search_photos(query , number_of_photos = 3 , page= 2):
    """
    Any query related to photos you must use this tool

    Args:

    query : the species name.
    number_of_photos :  number of photos to search and return.
    """
    
    taxon_id = get_taxa_id(query)

    base_url = "https://api.inaturalist.org/v1/observations"
    
    
    params = {
            "taxon_id": taxon_id,
            "per_page": number_of_photos,
            "page": page,
        }
    response = requests.get(base_url, params=params)
    if response.status_code != 200:
        print("Please Provide a valid query or species")
        # print(f"Failed to fetch data for page {page}: {response.status_code}")
    
    data = response.json()
    results = data.get("results", [])
    
    if not results:
        print(f"No more results found on page {page}.")

    photos2 = []
    for obs in results:
        photos = obs.get("photos", [])
        for photo in photos:
            image_url = photo.get("url")
            if image_url:
                photos2.append(image_url.replace("square" , "original"))

    if photos2:
        final_photos = random.sample(photos2 , int(number_of_photos))
        return prepare_content({"photos":final_photos})
    

    return {"error": "Please provide a valid query 2"}

## Tools With Agent

In [50]:
tools = [search , retriever_tool , search_photos]

In [110]:
# prompt = hub.pull("hwchase17/openai-functions-agent")

from langchain_core.prompts import ChatPromptTemplate

prompt = ChatPromptTemplate.from_messages([
  ("system", "You are a helpful assistant , if you don't find an answer in a tool then you should search online using the search tool , DO NOT ANSWER FROM YOUR OWN KNOWLEDGE"),
  ("placeholder", "{chat_history}"),
  ("human", "{input}"),
  ("placeholder", "{agent_scratchpad}"),
])

agent = create_tool_calling_agent(chat , tools , prompt)

agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True , return_intermediate_steps=True)

In [105]:
agent_executor.invoke({"input": "What does the green turtle eat?"})



> Entering new AgentExecutor chain...

Invoking: `Turtles_search` with `{'query': 'green turtle diet'}`


{"Green Turtle": {"supplementary_info": {"behavior_and_diet": "Green turtles are the only herbivorous species of sea turtle, primarily feeding on algae and seagrasses. Juveniles are initially carnivorous, eating fish eggs, mollusks, jellyfish, small invertebrates, worms, sponges, algae, and crustaceans. As they mature, they become omnivorous, and mostadults are strictly herbivorous. The East Pacific green turtle may consume more animal prey than other populations. Prior to moving to nearshore foraging areas, pelagic juveniles forage on plant and animal life found in oceanic drift communities"}}}

{"loggerhead turtle": {"supplementary_info": {"behavior_and_diet": "Loggerhead sea turtles are omnivorous, primarily feeding on bottom-dwelling invertebrates such as gastropods, bivalves, and decapods. They have a diverse diet that includes sponges, corals, sea pens, polychaete worms, se

{'input': 'What does the green turtle eat?',
 'output': 'Green turtles are the only herbivorous species of sea turtle. They primarily feed on algae and seagrasses. Juveniles start as carnivorous, consuming fish eggs, mollusks, jellyfish, small invertebrates, worms, sponges, algae, and crustaceans. As they mature, they become omnivorous, and most adults are strictly herbivorous. The East Pacific green turtle population may consume more animal prey than other populations. Prior to moving to nearshore foraging areas, pelagic juveniles forage on plant and animal life found in oceanic drift communities.',
 'intermediate_steps': [(ToolAgentAction(tool='Turtles_search', tool_input={'query': 'green turtle diet'}, log="\nInvoking: `Turtles_search` with `{'query': 'green turtle diet'}`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_ZWEWNQaaqVvLYpHN0ozCmImF', 'function': {'arguments': '{"query":"green turtle diet"}', 'name': 'Turtles_sea

## Tools , Memory With Agents

In [ ]:

agent_with_message_history = RunnableWithMessageHistory(
    agent_executor,
    lambda session_id: demo_ephemeral_chat_history,
    input_messages_key="input",
    history_messages_key="chat_history",
)


In [112]:
agent_with_trimming = (
    RunnablePassthrough.assign(messages_trimmed=trim_messages)
    | agent_with_message_history
)

In [ ]:
config = {"configurable" : {"session_id" : "raafat"}}


In [76]:
answer = agent_with_trimming.invoke({"input":"can you show me a photo for each of loggerhead turtle and leatherback turtle? also what does each of them eat?"} ,
                           config=config)



> Entering new AgentExecutor chain...

Invoking: `search_photos` with `{'query': 'loggerhead turtle', 'number_of_photos': 1}`


{
  "photos": [
    "https://inaturalist-open-data.s3.amazonaws.com/photos/468362945/original.jpg"
  ]
}
Invoking: `search_photos` with `{'query': 'leatherback turtle', 'number_of_photos': 1}`


{
  "photos": [
    "https://inaturalist-open-data.s3.amazonaws.com/photos/466510386/original.jpg"
  ]
}
Invoking: `Turtles_search` with `{'query': 'loggerhead turtle'}`


{"loggerhead turtle": {"supplementary_info": {"appearance": "The loggerhead turtle has a large head with powerful jaws and a slightly heart-shaped, reddish-brown carapace. The plastron (bottom shell) is pale yellow.adults weigh between 80 to 200 kg (180 to 440 lb) and measure 70 to 95 cm (28 to 37 in) in length, with the largest individuals reaching up to 545 kg (1,202 lb) and 213 cm (84 in). The neck and flippers are dull brown to reddish brown on top and medium to pale yellow on the sides and bot

In [ ]:
answer

# reasoning 

{'input': 'can you show me a photo for each of loggerhead turtle and leatherback turtle? also what does each of them eat?',
 'messages_trimmed': False,
 'chat_history': [HumanMessage(content='can you show me a photo for each of loggerhead turtle and leatherback turtle? also what does each of them eat?', additional_kwargs={}, response_metadata={}),
  AIMessage(content='Here is a photo of a Loggerhead Turtle:\n![Loggerhead Turtle](https://inaturalist-open-data.s3.amazonaws.com/photos/468316510/original.jpeg)\n\nThe Loggerhead turtle is the largest hard-shelled turtle found in oceans worldwide. They have a large head with powerful jaws and a slightly heart-shaped, reddish-brown carapace. Adults weigh around 135 kg and can live up to 67 years. Loggerhead turtles are omnivorous, mainly eating bottom-dwelling invertebrates.\n\nHere is a photo of a Leatherback Turtle:\n![Leatherback Turtle](https://inaturalist-open-data.s3.amazonaws.com/photos/464175323/original.jpeg)\n\nThe Leatherback turtl

In [114]:
answer = agent_with_trimming.invoke({"input":"What is meant by genus family?"} ,
                           config=config)



> Entering new AgentExecutor chain...
In biological classification, "genus" and "family" are two distinct levels of the taxonomic hierarchy used to categorize and organize living organisms.

1. **Genus**: This is a rank in the biological classification system that is above species and below family. A genus comprises one or more species that are closely related to each other. For example, in the scientific name *Homo sapiens*, "Homo" is the genus.

2. **Family**: This is a higher taxonomic rank than genus and includes one or more genera that share certain structural characteristics or common traits. For example, the family "Hominidae" includes the genus "Homo" along with other genera like "Pan" (chimpanzees).

Each level in this hierarchy helps scientists communicate more effectively about the diversity of life, drawing attention to shared characteristics and evolutionary relationships. The term "genus family" might be a confusion between these two separate categories.

> Finished cha

In [115]:
answer

{'input': 'What is meant by genus family?',
 'messages_trimmed': True,
 'chat_history': [HumanMessage(content='What is meant by genus family?', additional_kwargs={}, response_metadata={}),
  AIMessage(content='In biological classification, the term "genus family" is not a standard term or classification category. Each level in the taxonomic hierarchy (such as kingdom, phylum, class, order, family, genus, and species) represents a distinct rank with specific characteristics and relationships.\n\nA genus is a rank that comes below the family level and includes one or more closely related species. A family is a higher taxonomic rank than genus and consists of one or more genera that share common characteristics.\n\nIf you have a specific context or concept in mind where "genus family" is used, please provide more details so I can better understand and provide relevant information.', additional_kwargs={}, response_metadata={})],
 'output': 'In biological classification, "genus" and "family

#### Answer of the flatback sea turtle

In [85]:
answer["output"]

'The flatback sea turtle, scientifically known as Natator depressus, belongs to the following taxonomic classification:\n\n- Kingdom: Animalia\n- Phylum: Chordata\n- Class: Reptilia\n- Order: Testudines\n- Family: Cheloniidae\n- Genus: Natator\n- Species: N. depressus\n\nIf you have any more questions or need further information, feel free to ask!'

In [84]:
from langchain_core.output_parsers import MarkdownListOutputParser , StrOutputParser
import streamlit as st


# parser = MarkdownListOutputParser()
parser = StrOutputParser()
parser.parse(answer["output"])
# st.markdown(answer["output"])

'The flatback sea turtle, scientifically known as Natator depressus, belongs to the following taxonomic classification:\n\n- Kingdom: Animalia\n- Phylum: Chordata\n- Class: Reptilia\n- Order: Testudines\n- Family: Cheloniidae\n- Genus: Natator\n- Species: N. depressus\n\nIf you have any more questions or need further information, feel free to ask!'

'### Loggerhead Turtle:
\n-

 **Appearance**: The loggerhead turtle has a large head with powerful jaws and a slightly heart-shaped, reddish-brown carapace. Adults weigh between 80 to 200 kg (180 to 440 lb) and measure 70 to 95 cm (28 to 37 in) in length, with the largest individuals reaching up to 545 kg (1,202 lb) and 213 cm (84 in).\n- **Population**: Loggerheads are globally classified as a single species, with 10 biologically distinct Regional Management Units (RMUs) identified by nesting sites, genetic studies, and habitat use. Approximately 36,000-67,000 nesting females are estimated annually based on 200,000 clutches laid by all subpopulations.\n- **Habitat**: Loggerhead turtles nest on sandy beaches. Hatchlings emerge and migrate to the open ocean, utilizing nearshore and pelagic habitats during different life stages.\n- **Eating Habits**: Loggerhead turtles are omnivorous, mainly eating bottom-dwelling invertebrates.\n\n![Loggerhead Turtle](https://inaturalist-open-data.s3.amazonaws.com/photos/468197900/original.jpg)\n\n### Leatherback Turtle:\n- **Appearance**: Leatherback turtles have a distinctive appearance with a large, teardrop-shaped body and a smooth, leathery carapace. They lack a bony carapace and scales, instead having thick, leathery skin with embedded minuscule osteoderms. Adults average 1–1.75 meters in curved carapace length and weigh between 250 to 700 kg (550 to 1,540 lb).\n- **Population**: The leatherback sea turtle is the largest of all living turtles and the heaviest non-crocodilian reptile. It is the only living species in the genus Dermochelys and family Dermochelyidae. The species as a whole is considered vulnerable, and some of its subpopulations are critically endangered.\n- **Eating Habits**: Leatherback turtles have unique adaptations and impressive physical characteristics that make them well-suited for their extensive migrations and deep-sea diving.\n\n![Leatherback Turtle](https://inaturalist-open-data.s3.amazonaws.com/photos/466510386/original.jpg)\n\nIf you need more information or have any other questions, feel free to ask!'}

## Grader Agent

In [24]:
answer

{'input': 'can you show me a photo for each of loggerhead turtle and leatherback turtle? also what does each of them eat?',
 'messages_trimmed': False,
 'chat_history': [],
 'output': 'Here is a photo of a Loggerhead Turtle:\n![Loggerhead Turtle](https://inaturalist-open-data.s3.amazonaws.com/photos/468316510/original.jpeg)\n\nThe Loggerhead turtle is the largest hard-shelled turtle found in oceans worldwide. They have a large head with powerful jaws and a slightly heart-shaped, reddish-brown carapace. Adults weigh around 135 kg and can live up to 67 years. Loggerhead turtles are omnivorous, mainly eating bottom-dwelling invertebrates.\n\nHere is a photo of a Leatherback Turtle:\n![Leatherback Turtle](https://inaturalist-open-data.s3.amazonaws.com/photos/464175323/original.jpeg)\n\nThe Leatherback turtle is the largest of all living turtles and the heaviest non-crocodilian reptile. They have a distinctive appearance with a large, teardrop-shaped body and a smooth, leathery carapace. Le

In [51]:
import ast
@tool
def grader(agent_answer):
    """Grades the answer of the agent to either True or False.
    
    Args:
        - agent_answer: A dictionary containing the user's question and the previous agent's response.
    
    Returns:
        - True if the answer is relevant, False otherwise.
    """

    # Ensure agent_answer is a dictionary; if not, assume it's plain text and wrap it
    # if isinstance(agent_answer, str):
        # agent_answer = {"input": "Unknown Question", "output": agent_answer}

    prompt = ChatPromptTemplate.from_messages([
        ("system", "You are a grader that tells if the answer answers the question or not. Only return True or False."),
        ("human", "Question: {input}\nAnswer: {output}")
    ])

    grader_agent = prompt | chat

    grade = grader_agent.invoke({
        "input": agent_answer.get("question", "Unknown Question"),  # Default to avoid errors
        "output": agent_answer.get("answer", "No Answer Provided")
    })

    return grade.content

In [159]:
# prompt = hub.pull("hwchase17/openai-functions-agent")
import ast
from langchain_core.prompts import ChatPromptTemplate

prompt = ChatPromptTemplate.from_messages([
  ("system", """You are a helpful assistant. 
    - First, answer using retrieval if possible.
    - use the grader tool to check the quality of retrieval answer.
    - If the grader tool returns True, return the answer.
    - If the grader tool returns False, use the search tool instead.
    """),
  ("placeholder", "{chat_history}"),
  ("human", "{input}"),
  ("placeholder", "{agent_scratchpad}"),
])

tools = [search , search_photos , retriever_tool , grader]

agent = create_tool_calling_agent(chat , tools , prompt)

agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True , return_intermediate_steps=True )

In [160]:
test_1 = agent_executor.invoke({"input":"what is the hearing range of the flatback sea turtle"})



> Entering new AgentExecutor chain...

Invoking: `Turtles_search` with `{'query': 'hearing range of the flatback sea turtle'}`


The Australian flatback sea turtle (Natator depressus)[4] is a species of sea turtle in the family Cheloniidae. The species is endemic to the sandy beaches and shallow coastal waters of the Australian continental shelf. This turtle gets its common name from the fact that its shell has a flattened or lower dome than the other sea turtles. It can be olive green to grey with a cream underside. It averages from 76 to 96 cm (30 to 38 in) in carapace length and can weigh from 70 to 90 kg (150 to 200 lb). The hatchlings, when emerging from nests, are larger than other sea turtle hatchlings when they hatch.
The flatback turtle is listed by the IUCN Red List of Threatened Species as data deficient, meaning there is insufficient scientific information to determine its conservation status at this time.[1] It was previously listed as vulnerable in 1994.[5] It is not as

In [161]:
test_1

{'input': 'what is the hearing range of the flatback sea turtle',
 'output': "Flatback sea turtles, like other sea turtles, have a hearing range that is most sensitive to low-frequency sounds. Sea turtles generally hear best between 200 and 750 Hz and do not respond well to sounds above 1,000 Hz. While specific studies on the flatback sea turtle's hearing range are limited, it is likely similar to other sea turtles, which typically detect sounds in the range of 50 to 1,100 Hz, with their best sensitivity between 200 and 400 Hz.",
 'intermediate_steps': [(ToolAgentAction(tool='Turtles_search', tool_input={'query': 'hearing range of the flatback sea turtle'}, log="\nInvoking: `Turtles_search` with `{'query': 'hearing range of the flatback sea turtle'}`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_fXXybuCmHxEbvuUlWvAtMJCJ', 'function': {'arguments': '{"query":"hearing range of the flatback sea turtle"}', 'name': 'Turtles_search

In [158]:
test_1["output"]

'Sea turtles, including the flatback sea turtle, generally have a hearing range between 50 and 1,600 Hz. They are most sensitive to sounds below 1,000 Hz, with the best hearing between 200 and 750 Hz, and they do not respond well to sounds above 1,000 Hz.'

In [56]:
# from IPython.display import Image, display
# display(Image(agent.get_graph().draw_mermaid_png()))
# # agent_executor.ge

In [162]:
test_1["intermediate_steps"]

[(ToolAgentAction(tool='Turtles_search', tool_input={'query': 'hearing range of the flatback sea turtle'}, log="\nInvoking: `Turtles_search` with `{'query': 'hearing range of the flatback sea turtle'}`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_fXXybuCmHxEbvuUlWvAtMJCJ', 'function': {'arguments': '{"query":"hearing range of the flatback sea turtle"}', 'name': 'Turtles_search'}, 'type': 'function'}]}, response_metadata={'finish_reason': 'tool_calls', 'model_name': 'gpt-4o-2024-08-06', 'system_fingerprint': 'fp_4691090a87'}, id='run-b0a4c722-4220-4f57-b04c-928a4b14c8e2', tool_calls=[{'name': 'Turtles_search', 'args': {'query': 'hearing range of the flatback sea turtle'}, 'id': 'call_fXXybuCmHxEbvuUlWvAtMJCJ', 'type': 'tool_call'}], tool_call_chunks=[{'name': 'Turtles_search', 'args': '{"query":"hearing range of the flatback sea turtle"}', 'id': 'call_fXXybuCmHxEbvuUlWvAtMJCJ', 'index': 0, 'type': 'tool_call_chunk'}])], tool_

In [69]:
test_1

{'input': 'what is the taxa classification of the flatback sea turtle',
 'output': 'The flatback sea turtle belongs to the species Natator depressus in the family Cheloniidae.',
 'intermediate_steps': [(ToolAgentAction(tool='Turtles_search', tool_input={'query': 'flatback sea turtle'}, log="\nInvoking: `Turtles_search` with `{'query': 'flatback sea turtle'}`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_p9WNsq8nIDCmVxfs55haNsrN', 'function': {'arguments': '{"query":"flatback sea turtle"}', 'name': 'Turtles_search'}, 'type': 'function'}]}, response_metadata={'finish_reason': 'tool_calls', 'model_name': 'gpt-3.5-turbo-0125'}, id='run-ad20e98a-5933-44af-af7f-a2db16c0d290', tool_calls=[{'name': 'Turtles_search', 'args': {'query': 'flatback sea turtle'}, 'id': 'call_p9WNsq8nIDCmVxfs55haNsrN', 'type': 'tool_call'}], tool_call_chunks=[{'name': 'Turtles_search', 'args': '{"query":"flatback sea turtle"}', 'id': 'call_p9WNsq8nIDCmVxfs5

In [164]:

agent_with_message_history = RunnableWithMessageHistory(
    agent_executor,
    lambda session_id: demo_ephemeral_chat_history,
    input_messages_key="input",
    history_messages_key="chat_history",
)

In [165]:
agent_with_trimming_4o = (
    RunnablePassthrough.assign(messages_trimmed=trim_messages)
    | agent_with_message_history
)

In [167]:
agent_with_trimming_4o.invoke({"input":"can you show me a photo for each of loggerhead turtle and leatherback turtle? also what does each of them eat?"} ,
                           config=config)



> Entering new AgentExecutor chain...


RateLimitError: Error code: 429 - {'error': {'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.', 'type': 'insufficient_quota', 'param': None, 'code': 'insufficient_quota'}}

## Ollama , Gemini Embeddings

In [52]:
from langchain_google_genai import ChatGoogleGenerativeAI
from langchain_google_genai import GoogleGenerativeAIEmbeddings

google_api_key = "AIzaSyDlOfjWrjFzmtycyFBQtOLPof1cgoYRnMM"
llm_google = ChatGoogleGenerativeAI(model="gemini-1.5-flash" , temperature=0.7 , api_key= google_api_key)
embedding = GoogleGenerativeAIEmbeddings(model="models/embedding-001")

In [53]:
vector_store = FAISS.load_local("../../FinalEmbeddings/VectorStoreGemini",
                                     embedding,
                                     allow_dangerous_deserialization=True)

retriever_gemini = vector_store.as_retriever(kwargs={"k":5})

In [9]:
vector_store.similarity_search("what is the green turtle taxa classification?")

[Document(metadata={}, page_content='{"Green Turtle": {"Taxa": {"kingdom_name": "ANIMALIA", "phylum_name": "CHORDATA", "class_name": "REPTILIA", "order_name": "TESTUDINES", "family_name": "CHELONIIDAE", "genus_name": "Chelonia", "species_name": "mydas", "ara_kingdom_name": "\\u0627\\u0644\\u062d\\u064a\\u0648\\u0627\\u0646\\u0627\\u062a", "ara_phylum_name": "\\u0627\\u0644\\u062d\\u0628\\u0644\\u064a\\u0627\\u062a", "ara_class_name": "\\u0627\\u0644\\u0632\\u0648\\u0627\\u062d\\u0641", "ara_order_name": "\\u0627\\u0644\\u0633\\u0644\\u0627\\u062d\\u0641"}}}'),
 Document(metadata={}, page_content='{"Green Turtle": {"red_list_category": "Endangered at IUCN", "supplementary_info": {"lower_depth_limit": 200, "upper_depth_limit": 50, "life_style": {"characteristics": ["Aquatic", "Precocial", "Cosmopolitan", "Natatorial", "Terrestrial", "Oviparous", "Grazing"], "group_name": ["bale", "nest", "turn", "dole", "creep"]}, "weight": {"pounds": "150 to 400 pounds", "kg": "68 to 190 kilogram | lbs"

In [54]:
retriever_tool = create_retriever_tool(
    retriever_gemini,
    "Turtles_search",
    "Searches for information about the turtles , for any questions regarding the turtles you should use this tool!"
)

In [55]:
tools = [search , search_photos , retriever_tool  ]

In [15]:
# prompt = hub.pull("hwchase17/openai-functions-agent")
from langchain.agents import create_react_agent
from langchain_core.prompts import ChatPromptTemplate
# chat = ChatDeepSeek(model="MFDoom/deepseek-r1-tool-calling:8b",
#                     api_key=DEEPSEEK_API_KEY)
chat = ChatOllama(
    model="llama3.2"
)

chat = ChatOpenAI(
    model="llama3.2",
    api_key="ollama",
    base_url="http://localhost:11434/v1",
    temperature=0
)

chat = ChatGroq(
    model = "llama-3.1-70b-versatile",
    temperature=0,
    api_key= "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",
    max_retries=2,
)

prompt = ChatPromptTemplate.from_messages([
  ("system", "You are a helpful assistant , if you don't find an answer in a tool then you should search online using the search tool , DO NOT ANSWER FROM YOUR OWN KNOWLEDGE "
   "Only return the data"),
  ("placeholder", "{chat_history}"),
  ("human", "{input}"),
  ("placeholder", "{agent_scratchpad}"),
])

agent = create_tool_calling_agent(chat , tools , prompt)
# prompt = hub.pull("hwchase17/react")
# agent = create_react_agent(chat , tools , prompt)

agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True , return_intermediate_steps=True)

In [228]:
response = agent_executor.invoke({"input": "What is the hearing range of the flatback sea turtle?"})



> Entering new AgentExecutor chain...

Invoking: `Turtles_search` with `{'query': 'flatback sea turtle hearing range'}`


{"Flatback turtle": {"supplementary_info": {"cellularity": "multicellular , cellularity quality inhering in a bearer by virtue of the bearers consisting of more than one cell.", "nutrition_and_diet": "Flatback sea turtles have a carnivorous diet. They feed mostly on soft corals, sea cucumbers, shrimp, jellyfish, mollusks, and other invertebrates. They will also occasionally feed on seagrasses, even though they rarely eat vegetation.", "hearing range": ["500 Hz", "100 Hz"]}}}

{"Flatback turtle": {"supplementary_info": {"hearing threshold": ["35 dB"], "body mass": ["adult max : 93554.48 g"], "host_of": ["Neocosmospora falciformis (Carrion)"], "eat": ["cnidarians", "crustaceans", "mollusks", "echinoderms"], "geographic_distribution": ["Australia", "Papua New Guinea", "Indonesia"], "sexual system": "dioecious", "visual_system": "corneal system"}}}

{"Leatherback Turt

In [229]:
response["output"]

'The hearing range of the flatback sea turtle is between 100 Hz and 500 Hz.'

In [100]:
response

{'input': 'What is the full taxa classification of the green turtle ? what does the hawksbill turtle eat?',
 'output': 'Based on the search results, the full taxonomy classification of the Green Turtle is:\n\n* Kingdom: Animalia\n* Phylum: Chordata\n* Class: Reptilia\n* Order: Testudines\n* Family: Cheloniidae\n* Genus: Chelonia\n* Species: C. mydas\n\nAs for the diet of the Hawksbill Turtle, according to the search results, they primarily eat:\n\n* Sargassum fluitans\n* Stypopodium zonale\n* Padina gymnospora (Funnelweed)\n* Polysiphonia echinata\n* Solieria tenera\n\nNote that the diet of Hawksbill Turtles may vary depending on the location and availability of food sources.',
 'intermediate_steps': [(ToolAgentAction(tool='Turtles_search', tool_input={'query': 'green turtle taxonomy'}, log="\nInvoking: `Turtles_search` with `{'query': 'green turtle taxonomy'}`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_tnwjbnoh', 'functio

In [56]:
tools = [search , search_photos , retriever_tool , grader ]

In [232]:
# prompt = hub.pull("hwchase17/openai-functions-agent")
from langchain.agents import create_react_agent
from langchain_core.prompts import ChatPromptTemplate
# chat = ChatDeepSeek(model="MFDoom/deepseek-r1-tool-calling:8b",
#                     api_key=DEEPSEEK_API_KEY)
chat = ChatOllama(
    model="llama3.2"
)

chat = ChatOpenAI(
    model="llama3.2",
    api_key="ollama",
    base_url="http://localhost:11434/v1",
    temperature=0
)


prompt = ChatPromptTemplate.from_messages([
  ("system", """You are a helpful assistant. 
    - First, answer using retrieval if possible.
    - use the grader tool to check the quality of retrieval answer.
    - If the grader tool returns True, return the answer.
    - If the grader tool returns False, use the search tool instead.
    """),
  ("placeholder", "{chat_history}"),
  ("human", "{input}"),
  ("placeholder", "{agent_scratchpad}"),
])

agent = create_tool_calling_agent(chat , tools , prompt)
# prompt = hub.pull("hwchase17/react")
# agent = create_react_agent(chat , tools , prompt)

agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True , return_intermediate_steps=True)

In [237]:
response = agent_executor.invoke({"input": "What is the taxa classification of the kemp ridley sea turtle?"})



> Entering new AgentExecutor chain...

Invoking: `Turtles_search` with `{'query': 'kemp ridley sea turtle taxonomy'}`


{"Kemp ridley turtle": {"population": {"population_number": {"population_trend": "Increasing due to conservation efforts but remains vulnerable"}}, "use_and_trade": "The Kemp's Ridley is a protected species throughout its distribution in the Gulf of Mexico and Atlantic coastal waters of the U.S. There is no legal trade of this species"}}

{"Kemp ridley turtle": {"scientific_name": "Lepidochelys kempii", "TAXA": {"kingdom_name": "ANIMALIA", "phylum_name": "CHORDATA", "class_name": "REPTILIA", "order_name": "TESTUDINES", "family_name": "CHELONIIDAE", "genus_name": "Lepidochelys", "species_name": "kempii", "ara_kingdom_name": "\u0627\u0644\u062d\u064a\u0648\u0627\u0646\u0627\u062a", "ara_phylum_name": "\u0627\u0644\u062d\u0628\u0644\u064a\u0627\u062a", "ara_class_name": "\u0627\u0644\u0632\u0648\u0627\u062d\u0641"}}}

{"Kemp ridley turtle": {"distribution": {"vagrant_r

In [238]:
response


{'input': 'What is the taxa classification of the kemp ridley sea turtle?',
 'output': 'The taxonomic classification of the Kemp Ridley sea turtle is as follows:\n\n* Kingdom: Animalia\n* Phylum: Chordata\n* Class: Reptilia\n* Order: Testudines\n* Family: Cheloniidae\n* Genus: Lepidochelys\n* Species: L. kempii\n\nThis classification indicates that the Kemp Ridley sea turtle is a species of reptile, specifically a type of turtle, within the family Cheloniidae and genus Lepidochelys.',
 'intermediate_steps': [(ToolAgentAction(tool='Turtles_search', tool_input={'query': 'kemp ridley sea turtle taxonomy'}, log="\nInvoking: `Turtles_search` with `{'query': 'kemp ridley sea turtle taxonomy'}`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_kpmsu2hm', 'function': {'arguments': '{"query":"kemp ridley sea turtle taxonomy"}', 'name': 'Turtles_search'}, 'type': 'function'}]}, response_metadata={'finish_reason': 'stop', 'model_name': 'lla

### Anything

In [125]:

chat_with_tools = chat.bind_tools(tools=tools)

In [130]:
tools_model_response = chat_with_tools.invoke("what is the taxa classification of the green turtle , and show me one photo for it")

In [136]:
tools_model_response.additional_kwargs["tool_calls"]

[{'id': 'call_ryytcjf2',
  'function': {'arguments': '{"query":"green turtle taxonomy"}',
   'name': 'Turtles_search'},
  'type': 'function',
  'index': 0},
 {'id': 'call_onjfxord',
  'function': {'arguments': '{"number_of_photos":"1","page":"1","query":"green turtle"}',
   'name': 'search_photos'},
  'type': 'function',
  'index': 0}]

In [169]:
from langchain_core.messages import HumanMessage, ToolMessage

# Define the query and initialize messages
query = "what is the taxa classification of the green turtle"
messages = [HumanMessage(content=query)]

# Get the model's response with tool calls
ai_msg = chat_with_tools.invoke(messages)

# Iterate over the tool calls in the model's response
for tool in ai_msg.tool_calls:
    # Select the appropriate tool
    selected_tool = {"search_photos": search_photos, "Turtles_search": retriever}[tool["name"]]
    
    # Prepare the arguments for the tool
    if tool["name"] == "Turtles_search":
        tool_args = {"input": tool["args"]["query"]}  # Adjust for retriever's expected input
    else:
        tool_args = {"input" : tool["args"]}  # Pass arguments directly for other tools
    
    # Invoke the tool and get the response
    tool_response = selected_tool.invoke(**tool_args)
    
    # Append the tool's response as a ToolMessage
    messages.append(ToolMessage(content=str(tool_response), tool_call_id=tool["id"]))

# Invoke the model again with the updated messages (including tool responses)
final_response = chat_with_tools.invoke(messages)

# Print the final response
print(final_response)

content='It appears that you have a large JSON object containing information about parasites found on Green Sea Turtles. The structure of the data is not explicitly defined, but I can help you extract some insights from it.\n\nHere are a few observations:\n\n1. **Number of parasite species**: There are 132 different parasite species mentioned in the data.\n2. **Life stage distribution**: Some parasites have a specified life stage (e.g., "adult", "None"), while others do not. This might indicate that some parasites are still developing or have not yet reached adulthood.\n3. **Geographic distribution**: The data does not explicitly mention geographic locations, but it\'s possible that the parasite species mentioned are more common in certain regions.\n\nTo provide more specific insights, I would need to know what kind of analysis you\'re interested in performing on this data. Are you looking for:\n\n* A list of all unique parasite species?\n* An overview of the most common parasite speci

In [153]:
tools_model_response.tool_calls

[{'name': 'Turtles_search',
  'args': {'query': 'green turtle taxonomy'},
  'id': 'call_ryytcjf2',
  'type': 'tool_call'},
 {'name': 'search_photos',
  'args': {'number_of_photos': '1', 'page': '1', 'query': 'green turtle'},
  'id': 'call_onjfxord',
  'type': 'tool_call'}]

### Llama3.2 , Create React Agent

In [58]:
from langgraph.prebuilt import create_react_agent
from langchain_openai import ChatOpenAI
from langgraph.checkpoint.memory import MemorySaver

memory = MemorySaver()

chat = ChatOpenAI(
    model="qwen2.5",
    api_key="ollama",
    base_url="http://localhost:11434/v1",
    temperature=0
)

prompt = """You are a helpful assistant. 
    - First, answer using retrieval if possibl e (Turtles_search tool).
    - use the grader tool to check the quality of retrieval answer.
    - If the grader tool returns True, return the answer.
    - If the grader tool returns False, use the search tool instead.
    "This order must be followed strictly"
    """

research_agent = create_react_agent(chat , tools=tools , prompt= prompt , checkpointer=memory)

In [69]:
inputs = {"messages": [("user", "How does light pollution affect turtle?")]}
config = {"configurable": {"thread_id": "3"}}
research_agent.invoke(inputs , config=config)

{'messages': [HumanMessage(content='what is the taxa classification of the green turtle ? and can you show me one photo for it?', additional_kwargs={}, response_metadata={}, id='22e4f98a-fae1-4e0b-9244-c4140f58c324'),
  AIMessage(content='', additional_kwargs={'tool_calls': [{'id': 'call_mf7b3kk8', 'function': {'arguments': '{"query":"green turtle taxa classification"}', 'name': 'Turtles_search'}, 'type': 'function', 'index': 0}, {'id': 'call_yiclb15h', 'function': {'arguments': '{"number_of_photos":1,"query":"green turtle"}', 'name': 'search_photos'}, 'type': 'function', 'index': 0}], 'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 53, 'prompt_tokens': 525, 'total_tokens': 578, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'qwen2.5', 'system_fingerprint': 'fp_ollama', 'finish_reason': 'tool_calls', 'logprobs': None}, id='run-112bd9d6-8839-45ce-844c-42554441c16f-0', tool_calls=[{'name': 'Turtles_search', 'args': {'query': 'gr

In [ ]:
memory.

AttributeError: 'MemorySaver' object has no attribute 'load'


- Question : what is the taxa classification of the green turtle and can you show me one photo for it 

The taxonomic classification of the Green Turtle is as follows: 

    Kingdom: Animalia
    Phylum: Chordata 
    Class: Reptilia 
    Order: Testudines
    Family: Cheloniidae
    Genus: Chelonia
    Species: C. mydas

Here's a photo of a Green Turtle :  https://static.inaturalist.org/photos/469029661/original.jpeg 

-Note: The image is from the Inaturalist platform and may not be available in all regions due to copyright restrictions."

### Qwen2.5 , create tool calling agent

### Customizing Search Tavily Tool

In [ ]:
from langchain.tools import  Tool
tavily_tool = TavilySearchResults(max_results=3)

def search_tool_wrapper(query):
    return tavily_tool.invoke(query)

search_tool_wrapper("  does the green turtle eat?")

search_tool = Tool(name="search_tool_json",
                   func=search_tool_wrapper,
                   description="used for any search queries")

tools = [tavily_tool , search_photos , retriever_tool , grader]

In [ ]:
from langchain_core.prompts import  SystemMessagePromptTemplate , ChatPromptTemplate
memory = MemorySaver()

chat = ChatOpenAI(
    model="llama3.2",
    api_key="ollama",
    base_url="http://localhost:11434/v1",
    temperature=0
)

instructions = """You are a helpful assistant. 
    - First, answer using retrieval if possible (Turtles_search tool).
    - use the grader tool to check the quality of retrieval answer.
    - If the grader tool returns True, return the answer.
    - If the grader tool returns False, use the search tool instead.
    """

prompt = ChatPromptTemplate.from_messages([SystemMessagePromptTemplate.from_template(instructions)])

research_agent = create_react_agent(chat , tools=tools , prompt= instructions , checkpointer=memory)

In [69]:
inputs = {"messages": [("user", "Who is lionel messi?")]}
config = {"configurable": {"thread_id": "1"}}
research_agent.invoke(inputs , config=config)

{'messages': [HumanMessage(content='Who is lionel messi?', additional_kwargs={}, response_metadata={}, id='1db127da-6c92-4507-ac5f-c5a8d0f8a707'),
  AIMessage(content='', additional_kwargs={'tool_calls': [{'id': 'call_g7e9v17n', 'function': {'arguments': '{"query":"lionel messi"}', 'name': 'Turtles_search'}, 'type': 'function', 'index': 0}], 'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 21, 'prompt_tokens': 508, 'total_tokens': 529, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'llama3.2', 'system_fingerprint': 'fp_ollama', 'finish_reason': 'tool_calls', 'logprobs': None}, id='run-286d0b4c-9462-4937-bcca-7159a09b86d9-0', tool_calls=[{'name': 'Turtles_search', 'args': {'query': 'lionel messi'}, 'id': 'call_g7e9v17n', 'type': 'tool_call'}], usage_metadata={'input_tokens': 508, 'output_tokens': 21, 'total_tokens': 529, 'input_token_details': {}, 'output_token_details': {}}),
  ToolMessage(content='{"Green Turtle": {"maps": {"U

### Groq

In [12]:
chat = ChatGroq(
    model = "llama-3.3-70b-versatile",
    temperature=0,
    api_key= "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",
    max_retries=2,
)

prompt = ChatPromptTemplate.from_messages([
  ("system", "You are a helpful assistant , if you don't find an answer in a tool then you should search online using the search tool , DO NOT ANSWER FROM YOUR OWN KNOWLEDGE "
   "Only return the data"),
  ("placeholder", "{chat_history}"),
  ("human", "{input}"),
  ("placeholder", "{agent_scratchpad}"),
])

agent = create_tool_calling_agent(chat , tools , prompt)
# prompt = hub.pull("hwchase17/react")
# agent = create_react_agent(chat , tools , prompt)

agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True , return_intermediate_steps=True)

In [13]:
response = agent_executor.invoke({"input": "What is the taxa classification of the green turtle? and can you show me one photo for it?"})



> Entering new AgentExecutor chain...

Invoking: `Turtles_search` with `{'query': 'green turtle taxa classification'}`


{"Green Turtle": {"Taxa": {"kingdom_name": "ANIMALIA", "phylum_name": "CHORDATA", "class_name": "REPTILIA", "order_name": "TESTUDINES", "family_name": "CHELONIIDAE", "genus_name": "Chelonia", "species_name": "mydas", "ara_kingdom_name": "\u0627\u0644\u062d\u064a\u0648\u0627\u0646\u0627\u062a", "ara_phylum_name": "\u0627\u0644\u062d\u0628\u0644\u064a\u0627\u062a", "ara_class_name": "\u0627\u0644\u0632\u0648\u0627\u062d\u0641", "ara_order_name": "\u0627\u0644\u0633\u0644\u0627\u062d\u0641"}}}

{"Green Turtle": {"documentation": {"range": {"oceans": {"atlantic": ["Eastern Central", "Northeast", "Northwest", "Southeast", "Southwest", "Western Central"], "indian": ["Eastern", "Western"], "pacific": ["Eastern Central", "Northwest", "Southwest", "Western Central"]}, "distribution": {"regions": "Circumglobal", "waters": "Tropical and subtropical"}}}}}

{"Green Turtle": {"s

In [23]:
response["output"]

'The taxa classification of the green turtle is as follows:\n- Kingdom: ANIMALIA \n- Phylum: CHORDATA \n- Class: REPTILIA \n- Order: TESTUDINES \n- Family: CHELONIIDAE \n- Genus: Chelonia \n- Species: mydas \n\nA photo of the green turtle can be found at: \nhttps://inaturalist-open-data.s3.amazonaws.com/photos/469659497/original.jpeg'

In [15]:

agent_with_message_history = RunnableWithMessageHistory(
    agent_executor,
    lambda session_id: demo_ephemeral_chat_history,
    input_messages_key="input",
    history_messages_key="chat_history",
)

In [32]:
agent_with_trimming_groq = (
    RunnablePassthrough.assign(messages_trimmed=trim_messages)
    | agent_with_message_history
)

In [36]:
config = {"configurable" : {"session_id" : "raafat"}}
agent_with_trimming_groq.invoke({"input":"show me some photos for it"} ,
                           config=config)



> Entering new AgentExecutor chain...

Invoking: `search_photos` with `{'query': "Kemp's ridley turtle", 'number_of_photos': 10}`


{'total_results': 756, 'page': 1, 'per_page': 10, 'results': [{'quality_grade': 'research', 'time_observed_at': '2023-07-01T14:12:00-05:00', 'taxon_geoprivacy': 'obscured', 'annotations': [{'uuid': 'eafa854d-ec49-49e1-8b8a-c5ea21c5b12c', 'controlled_attribute_id': 1, 'controlled_value_id': 8, 'concatenated_attr_val': '1|8', 'user_id': 6934023, 'votes': [], 'user': {'id': 6934023, 'login': 'dark_l0rd2', 'spam': False, 'suspended': False, 'created_at': '2023-05-23T19:58:18+00:00', 'login_autocomplete': 'dark_l0rd2', 'login_exact': 'dark_l0rd2', 'name': 'Matthew', 'name_autocomplete': 'Matthew', 'orcid': None, 'icon': 'https://static.inaturalist.org/attachments/users/icons/6934023/eb318459e1491878508b352e0c880949-thumb.jpg?1738370779', 'observations_count': 2079, 'identifications_count': 136936, 'journal_posts_count': 0, 'activity_count': 139015, 'species_c

{'input': 'show me some photos for it',
 'messages_trimmed': True,
 'chat_history': [HumanMessage(content='What does it eat?', additional_kwargs={}, response_metadata={}),
  AIMessage(content="The Kemp's ridley turtle feeds on mollusks, crustaceans, jellyfish, fish, algae or seaweed, and sea urchins. Juveniles primarily feed on crabs. Some specific examples of their diet include:\n\n* Callinectes sapidus (blue crab)\n* Penaeus setiferus (Bluetailed shrimp)\n* Arenaeus cribrarius (speckled swimming crab)\n* Neverita duplicata (lobed moonsnail)\n* Menippe adina (Gulf stone crab)\n\nThey are omnivores and their diet changes as they mature.", additional_kwargs={}, response_metadata={})],
 'output': "The Kemp's ridley turtle feeds on mollusks, crustaceans, jellyfish, fish, algae or seaweed, and sea urchins. Juveniles primarily feed on crabs. Some specific examples of their diet include:\n\n* Callinectes sapidus (blue crab)\n* Penaeus setiferus (Bluetailed shrimp)\n* Arenaeus cribrarius (spe

In [33]:
config = {"configurable" : {"session_id" : "raafat"}}
agent_with_trimming_groq.invoke({"input":"can you show me a photo for each of loggerhead turtle and leatherback turtle? also what does each of them eat?"} ,
                           config=config)



> Entering new AgentExecutor chain...

Invoking: `search_photos` with `{'query': 'loggerhead turtle', 'number_of_photos': 1}`


{'total_results': 19076, 'page': 1, 'per_page': 1, 'results': [{'quality_grade': 'research', 'time_observed_at': '2025-02-05T09:44:00+05:00', 'taxon_geoprivacy': 'obscured', 'annotations': [], 'uuid': '5846cf67-6b7a-4a2d-89bb-730b5bfd4518', 'observed_on_details': {'date': '2025-02-05', 'day': 5, 'month': 2, 'year': 2025, 'hour': 9, 'week': 6}, 'id': 261391687, 'cached_votes_total': 0, 'identifications_most_agree': True, 'created_at_details': {'date': '2025-02-12', 'day': 12, 'month': 2, 'year': 2025, 'hour': 4, 'week': 7}, 'species_guess': 'kareta pravá', 'identifications_most_disagree': False, 'tags': [], 'positional_accuracy': 689, 'comments_count': 0, 'site_id': 1, 'created_time_zone': 'Indian/Maldives', 'license_code': 'cc-by-nc', 'observed_time_zone': 'Indian/Maldives', 'quality_metrics': [], 'public_positional_accuracy': 31418, 'reviewed_by': [1847050,

APIError: Failed to call a function. Please adjust your prompt. See 'failed_generation' for more details.

In [42]:
answer = agent_executor.invoke({"input":"what is the hearing range of the flatback sea turtle?"})



> Entering new AgentExecutor chain...

Invoking: `Turtles_search` with `{'query': 'flatback sea turtle hearing range'}`


{"Flatback turtle": {"supplementary_info": {"cellularity": "multicellular , cellularity quality inhering in a bearer by virtue of the bearers consisting of more than one cell.", "nutrition_and_diet": "Flatback sea turtles have a carnivorous diet. They feed mostly on soft corals, sea cucumbers, shrimp, jellyfish, mollusks, and other invertebrates. They will also occasionally feed on seagrasses, even though they rarely eat vegetation.", "hearing range": ["500 Hz", "100 Hz"]}}}

{"Flatback turtle": {"supplementary_info": {"hearing threshold": ["35 dB"], "body mass": ["adult max : 93554.48 g"], "host_of": ["Neocosmospora falciformis (Carrion)"], "eat": ["cnidarians", "crustaceans", "mollusks", "echinoderms"], "geographic_distribution": ["Australia", "Papua New Guinea", "Indonesia"], "sexual system": "dioecious", "visual_system": "corneal system"}}}

{"Leatherback Turt

In [43]:
answer["output"]

'The hearing range of the flatback sea turtle is between 100 Hz and 500 Hz, with a hearing WaitForSeconds of 35 dB.'

In [44]:
answer["intermediate_steps"]

[(ToolAgentAction(tool='Turtles_search', tool_input={'query': 'flatback sea turtle hearing range'}, log="\nInvoking: `Turtles_search` with `{'query': 'flatback sea turtle hearing range'}`\n\n\n", message_log=[AIMessageChunk(content='', additional_kwargs={'tool_calls': [{'index': 0, 'id': 'call_gef5', 'function': {'arguments': '{"query": "flatback sea turtle hearing range"}', 'name': 'Turtles_search'}, 'type': 'function'}]}, response_metadata={'finish_reason': 'tool_calls'}, id='run-95124668-1466-4c99-a59c-d70be86a5d1c', tool_calls=[{'name': 'Turtles_search', 'args': {'query': 'flatback sea turtle hearing range'}, 'id': 'call_gef5', 'type': 'tool_call'}], usage_metadata={'input_tokens': 531, 'output_tokens': 20, 'total_tokens': 551}, tool_call_chunks=[{'name': 'Turtles_search', 'args': '{"query": "flatback sea turtle hearing range"}', 'id': 'call_gef5', 'index': 0, 'type': 'tool_call_chunk'}])], tool_call_id='call_gef5'),
  '{"Flatback turtle": {"supplementary_info": {"cellularity": "mu

### Redis Test

In [ ]:
from langchain.memory.buffer import ConversationBufferMemory 
from langchain_community.chat_message_histories import RedisChatMessageHistory


In [5]:
REDIS_URL = "redis://localhost:6379/0"
def get_message_history(session_id: str) -> RedisChatMessageHistory:
    return RedisChatMessageHistory(session_id, url=REDIS_URL)

In [9]:
from langchain_openai import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate , MessagesPlaceholder
from langchain_core.runnables.history import RunnableWithMessageHistory

model = ChatOpenAI(model="qwen2.5" , api_key="ollama" , base_url="http://localhost:11434/v1")
prompt = ChatPromptTemplate.from_messages(
    [
        (
            "system",
            "You're an assistant who's good at {ability}. Respond in 20 words or fewer",
        ),
        MessagesPlaceholder(variable_name="history"),
        ("human", "{input}"),
    ]
)
runnable = prompt | model

with_message_history = RunnableWithMessageHistory(
    runnable,
    get_message_history,
    input_messages_key="input",
    history_messages_key="history",
)

In [41]:
with_message_history.invoke(
    {"ability": "math", "input": "repeat what you just said?"},
    config={"configurable": {"session_id": "nabil"}},
)

AIMessage(content="What does cosine mean? I don't know.", additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 11, 'prompt_tokens': 82, 'total_tokens': 93, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'qwen2.5', 'system_fingerprint': 'fp_ollama', 'finish_reason': 'stop', 'logprobs': None}, id='run-7392e510-27fd-493a-a05c-7e57b52cf8c6-0', usage_metadata={'input_tokens': 82, 'output_tokens': 11, 'total_tokens': 93, 'input_token_details': {}, 'output_token_details': {}})

In [12]:
from langchain.schema import AIMessage
get_message_history("foobar").messages[-1] = AIMessage(content = "I dont know")

In [42]:
foobar_session = get_message_history("nabil").messages
foobar_session

[HumanMessage(content='What does cosine mean?', additional_kwargs={}, response_metadata={}),
 AIMessage(content="I don't know", additional_kwargs={}, response_metadata={}),
 HumanMessage(content='repeat what you just said?', additional_kwargs={}, response_metadata={}),
 AIMessage(content="What does cosine mean? I don't know.", additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 11, 'prompt_tokens': 56, 'total_tokens': 67, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'qwen2.5', 'system_fingerprint': 'fp_ollama', 'finish_reason': 'stop', 'logprobs': None}, id='run-e38d864c-2a26-4bae-abb3-8b155b912742-0', usage_metadata={'input_tokens': 56, 'output_tokens': 11, 'total_tokens': 67, 'input_token_details': {}, 'output_token_details': {}}),
 HumanMessage(content='repeat what you just said?', additional_kwargs={}, response_metadata={}),
 AIMessage(content="What does cosine mean? I don't know.", additional_kwargs={'ref

In [17]:
foobar_session[-1].content = "I dont know"

In [35]:
chat_history = get_message_history("nabil")

if chat_history.messages:
    # Get all messages except the last one
    messages_except_last = chat_history.messages[:-1]
    print(f"Before clearing")    
    print(chat_history.messages)    
    # Clear the entire chat history in Redis
    chat_history.clear()
    
    # Add back all messages except the last one
    for message in messages_except_last:
        chat_history.add_message(message)
    
    # Add the new modified message
    chat_history.add_message(AIMessage(content="I don't know"))

Before clearing
[HumanMessage(content='What does cosine mean?', additional_kwargs={}, response_metadata={}), AIMessage(content='Cosine is a trigonometric function, the x-coordinate of a point on the unit circle.', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 21, 'prompt_tokens': 36, 'total_tokens': 57, 'completion_tokens_details': None, 'prompt_tokens_details': None}, 'model_name': 'qwen2.5', 'system_fingerprint': 'fp_ollama', 'finish_reason': 'stop', 'logprobs': None}, id='run-0822dbf3-751c-4ccc-a8c8-6fb371ca7c81-0', usage_metadata={'input_tokens': 36, 'output_tokens': 21, 'total_tokens': 57, 'input_token_details': {}, 'output_token_details': {}})]


In [36]:
get_message_history("nabil").messages

[HumanMessage(content='What does cosine mean?', additional_kwargs={}, response_metadata={}),
 AIMessage(content="I don't know", additional_kwargs={}, response_metadata={})]

### Putting it all together